src/terminal/parser/stateMachine.cpp

```diff
@@ -1634,21 +1640,52 @@ void StateMachine::_EventDcsPassThrough(const wchar_t wch)
 }
 
 // Routine Description:
-// - Handle the two-character termination of a DCS sequence.
+// - Handle SOS/PM/APC string.
 //   Events in this state will:
-//   1. Enter ground on a string terminator
-//   2. Pass on everything else as the start of a regular escape sequence
+//   1. If we see a ESC, enter the SosPmApcTermination state.
+//   2. Ignore everything else.
 // Arguments:
 // - wch - Character that triggered the event
 // Return Value:
 // - <none>
-void StateMachine::_EventDcsTermination(const wchar_t wch)
+void StateMachine::_EventSosPmApcString(const wchar_t wch) noexcept
 {
-    _trace.TraceOnEvent(L"DcsTermination");
+    _trace.TraceOnEvent(L"SosPmApcString");
+    if (_isEscape(wch))
+    {
+        _EnterSosPmApcTermination();
+    }
+    else
+    {
+        _ActionIgnore();
+    }
+}
 
+// Routine Description:
+// - Handle "Variable Length String" termination.
+//   Events in this state will:
+//   1. Trigger the corresponding action and enter ground if we see a string terminator,
+//   2. Otherwise treat this as a normal escape character event.
+// Arguments:
+// - wch - Character that triggered the event
+// Return Value:
+// - <none>
+void StateMachine::_EventVariableLengthStringTermination(const wchar_t wch)
+{
     if (_isStringTerminatorIndicator(wch))
     {
-        // TODO: The Dcs sequence has successfully terminated. This is where we'd be dispatching the DCS command.
+        if (_state == VTStates::OscTermination)
+        {
+            _ActionOscDispatch(wch);
+        }
+        if (_state == VTStates::DcsTermination)

```

Comment:
should these be `else if`s? I get that `_Action*Dispatch` _shouldn't_ change the state, but still...

(this is a nit)

Reply:
I'll fix those nits in my next PR for #7316 if it's OK with you.  Thanks Mike, as always.

Reply:
I'm having some trouble with the fix for #7316 where the performance of DCS passthrough is not good enough. So I won't be able to file a new PR soon. I think we can ship this PR first and see where it takes us.